### PR TITLE
[RLlib] OPE (off policy estimator) API.

### PR DESCRIPTION
**Pull request by sven1977**
OPE (off policy estimator) API. The existing OffPolicyEstimator (API) class can now be sub-classed by users and these new sub-classes can be specified in the `input_evaluation` config field to allow for any custom evaluators to be used. Before this change, only the hard-coded strings "is" (pointing to the ImportanceSampling class) and "wis" (pointing to the WeightedImportanceSampling class) were allowed.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/trainer_config.py [hunk 1]
```diff
@@ -170,7 +172,8 @@ def __init__(self, trainer_class=None):
         self.input_ = "sampler"
         self.input_config = {}
         self.actions_in_input_normalized = False
-        self.input_evaluation = ["is", "wis"]
+        self.input_evaluation = [ImportanceSamplingEstimator,
```
Comment (maxpumperla) [inline rllib/agents/trainer_config.py:175]: loving this! the more cryptic acronyms we can replace, the better. Having said that, maybe we can settle for a bit of a compromise here in terms of brevity. My suggestion would be:

```python
from ray.rllib.offline import ImportanceSampling
from ray.rllib.offline import WeightedImportanceSampling
```

i.e. import everything in `offline` and drop the `Estimator`. wdyt?

File: rllib/agents/trainer_config.py [hunk 2]
```diff
@@ -866,10 +869,11 @@ def offline_data(
             input_evaluation: Specify how to evaluate the current policy. This only has
                 an effect when reading offline experiences ("input" is not "sampler").
                 Available options:
-                - "wis": the weighted step-wise importance sampling estimator.
-                - "is": the step-wise importance sampling estimator.
-                - "simulation": run the environment in the background, but use
+                - "simulation": Run the environment in the background, but use
                 this data for evaluation only and not for learning.
+                - Any subclass of OffPolicyEstimator, e.g.
+                ray.rllib.offline.is_estimator::ImportanceSamplingEstimator or your own
```
Comment (maxpumperla) [inline rllib/agents/trainer_config.py:875]: Can we document the available options, though? what are is and wis and when would I use each?

File: rllib/evaluation/rollout_worker.py [hunk 3]
```diff
@@ -345,11 +345,12 @@ def __init__(
...
-                reading offline data. The possible values include:
-                - "is": the step-wise importance sampling estimator.
-                - "wis": the weighted step-wise is estimator.
-                - "simulation": run the environment in the background, but
-                use this data for evaluation only and never for learning.
+                reading offline data. Available options:
+                - "simulation": Run the environment in the background, but use
+                this data for evaluation only and not for learning.
+                - Any subclass of OffPolicyEstimator, e.g.
+                ray.rllib.offline.is_estimator::ImportanceSamplingEstimator or your own
```
Comment (maxpumperla) [inline rllib/evaluation/rollout_worker.py:352]: same as above

**Review by avnishn — APPROVED** [review]

File: rllib/agents/trainer_config.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/trainer_config.py]: changed

**Review by gjoliver — COMMENTED** [review]
I see double-colon used quite a few times. what does that mean actually?

File: rllib/agents/trainer.py [hunk 4]
```diff
@@ -558,11 +562,15 @@
...
-    #  - "simulation": run the environment in the background, but use
+    #  - "simulation": Run the environment in the background, but use
     #    this data for evaluation only and not for learning.
-    "input_evaluation": ["is", "wis"],
+    #  - Any subclass of OffPolicyEstimator, e.g.
+    #    ray.rllib.offline.estimators.is::ImportanceSampling or your own custom
+    #    subclass.
+    "input_evaluation": [
+        ImportanceSampling,
+        WeightedImportanceSampling,
```
Comment (gjoliver) [inline rllib/agents/trainer.py:572]: also just curious, why do we run both by default.

Comment (avnishn) [reply rllib/agents/trainer.py:572]: Oh is that what this means? Same question then... why not run 1, giving users the option which one that they want to run?

Comment (sven1977) [reply rllib/agents/trainer.py:572]: Short answer: It's always been this way :D 

Happy to change this default behavior, but let's do this in a different PR and only if we have a good reason to break this (maybe some users rely on both being there and don't even know, why these are always there).

**Pull request merged**